export enum AuthApiErrorMessage {
  "bad_json" = "잘못된 요청 형식입니다.", // 잘못된 JSON 형식
  "bad_jwt" = "유효하지 않은 인증 토큰입니다.", // 잘못된 JWT 토큰
  "conflict" = "동시성 충돌이 발생했습니다. 다시 시도해주세요.", // 동시성 충돌
  "email_exists" = "이미 존재하는 이메일입니다.", // 이메일이 이미 존재함
  "invalid_credentials" = "잘못된 이메일 또는 비밀번호입니다.", // 잘못된 이메일 혹은 비밀번호
  "email_provider_disabled" = "이메일 로그인이 비활성화되었습니다.", // 이메일 로그인 비활성화,
  "no_authorization" = "인증 헤더가 없습니다.", // 인증 헤더 없음
  "not_admin" = "관리자 권한이 없습니다.", // 관리자 권한 없음
  "user_already_exists" = "이미 존재하는 유저입니다.", // 이미 유저가 존재함
}

// Storage API 오류 메시지 매핑: Error가 code를 제공하지 않음
export const StorageApiErrorMessage = {
  404: "요청하신 리소스를 찾을 수 없습니다.", // 리소스 없음
  401: "유효하지 않은 JWT 토큰입니다.", // 잘못된 JWT 토큰
  400: "잘못된 요청 형식입니다.", // 잘못된 요청 형식
  403: "권한이 없습니다.", // 권한 없음
  409: "리소스가 이미 존재합니다.", // 리소스 충돌
  429: "요청이 너무 많습니다. 잠시 후 다시 시도해주세요.", // 너무 많은 요청
  544: "데이터베이스에 요청이 많아 효율적으로 처리할 수 없습니다. 잠시 후 다시 시도해주세요.", // DB 요청 과부하
  500: "서버 내부 오류가 발생했습니다.", // 서버 내부 오류
};

// 데이터베이스 API 오류 메시지
export const DatabaseApiErrorMessage = {
  "08": "데이터베이스에 연결할 수 없습니다. 잠시 후 다시 시도해주세요.", // 연결 오류,
  "0L": "이 작업을 수행할 권한이 없습니다.", // 권한 없음
  "0P": "잘못된 권한 설정입니다.", // 잘못된 권한
  "23503": "연관된 데이터 때문에 작업을 완료할 수 없습니다.", // foreign 데이터 오류
  "23505": "이미 존재하는 데이터입니다.", // 중복 데이터
  "28": "인증 정보가 올바르지 않습니다. 다시 로그인해 주세요.", // 인증 정보 오류
  "53": "서버에 리소스가 부족합니다. 잠시 후 시도하세요.", // 리소스 부족
  "42883": "정의되지 않은 Function입니다.", // 정의되지 않은 Function
  "42P01": "정의되지 않은 테이블입니다.", // 잘못된 테이블
  "42501": "인증되지 않았거나 권한이 없는 사용자입니다.", // 인증되지 않음 or 권한 없음
};

export type ErrorReturn = {
  name: "validation" | "server" | "network" | "unexpected";
  message: string;
  code?: string;
  errors?: Record<string, string[]>;
};
